File: misc/gemini_competition/static_content_pipeline/src/generators/chapterGenerator.py
import json
import google.generativeai as genai
from config.constants import LIGHT_BLUE, MATTE_GREEN, COOL_PINK, RESET
from config.courses.eng_de.constants import AVOID_JSON
from config.courses.eng_de.constants import COURSE, ORIGIN, LANGUAGE
import os


COURSE = "eng_de"
ORIGIN = "english"
LANGUAGE = "german"

class ChapterGenerator():

    example = ""

    # Load the JSON file from the path
    def load_json(self, path):
        with open(path, 'r') as file:
            data = json.load(file)
        return data

    def __init__(self):
        current_dir = os.path.dirname(os.path.realpath(__file__))
        self.json_type_structure = self.load_json(os.path.join(current_dir, 'exercise_types.json'))

    def create_system_instruction(self, module, section, level):
        return f"""
          **Your Role:**

          - You are a virtual language teacher for students of {ORIGIN} learning {LANGUAGE} at level {level}.
          - Your goal is to create a JSON file with one exercise chapter for module {module} section {section} at level {level}.
              
          **Your task is to generate a New Chapter based on the given information**
          """

    def generate_chapter(self, module, section, type, chapter_number, level, chapter_description):
        match type:
            case 0:
                example = self.json_type_structure['chapters'][0]
                self.example = example
            case 1:
                example = self.json_type_structure['chapters'][1]
                self.example = example
            case 2:
                example = self.json_type_structure['chapters'][2]
                self.example = example
            case 3:
                example = self.json_type_structure['chapters'][3]
                self.example = example
            case 4:
                example = self.json_type_structure['chapters'][4]
                self.example = example
            case 5:
                example = self.json_type_structure['chapters'][5]
                self.example = example
            case 6:
                example = self.json_type_structure['chapters'][6]
                self.example = example
            case 7:
                example = self.json_type_structure['chapters'][7]
                self.example = example
            case _:
                print("Invalid type")

        
        instruction = self.create_system_instruction(module, section, level)
        model = genai.GenerativeModel(
            model_name="gemini-1.5-pro",
            system_instruction=instruction,
            generation_config={"response_mime_type": "application/json"}
        )
        response = model.generate_content(
            f"Generate a chapter of type {type} for module {module} section {section} with exercises for students of {ORIGIN} learning {LANGUAGE} at level {level}. The chapter number is {chapter_number}. The chapter description and content is as follows: {chapter_description}. Here is an exmaple for the JSON structure: {self.example}. The chapter content should match the amount of the examplary content. Please avoid the mistakes mentioned in the JSON structure: {AVOID_JSON}"
        )
        
        if response.text is None:
            # TODO if this ever happends need to implement retry logic
            print(f"{COOL_PINK}No response from the Gemini API. Exiting.{RESET}")
            return
    
        try:
            response_json = json.loads(response.text)
            print(response.text)
        except json.JSONDecodeError as e:
            print(f"{COOL_PINK}Failed to parse JSON: {e}{RESET}")
            print("Response text was:", response.text)
            print(f"{LIGHT_BLUE}Reasking for the correct JSON format...{RESET}")
        
        self.save_json_to_file(response_json, module=module, section=section, level=level, chapter=chapter_number)
        
        return response_json
    

    def save_json_to_file(self, response, module, section, level, chapter):
        if response is None:
            print("No valid JSON to save. Exiting.")
            return

        directory_path = f"../../modules/{COURSE}_{level}/module{module}/sections/section{section}"
        os.makedirs(directory_path, exist_ok=True)
        file_name = f"chapter_{chapter}.json"
        file_path = os.path.join(directory_path, file_name)
        
        try:
            with open(file_path, 'w') as json_file:
                json.dump(response, json_file, indent=4)
            print(f"{MATTE_GREEN}Response saved to {file_path}{RESET}")
        except Exception as e:
            print(f"{COOL_PINK}Failed to save the response to the file: {e}{RESET}")
        

# This is only used when the script is run directly
if __name__ == "__main__":
    module = 1
    section = 2
    type = 5
    chapter_number = 1
    level = "A1"

    chapter_generator = ChapterGenerator()
    response = chapter_generator.generate_chapter(module=module, section=section, type=type, chapter_number=chapter_number, level=level)

File: cloud_functions_node/functions/chapterGenerator.js
const {logger} = require("firebase-functions");
const {VertexAI} = require("@google-cloud/vertexai");

/**
 * Generates a chapter exercise for language learning.
 *
 * @param {string} origin - Origin language.
 * @param {string} target - Target Language.
 */
class ChapterGenerator {
  /**
 * Create a new ChapterGenerator.
 *
 * @param {string} origin - Origin language.
 * @param {string} language - Target Language.
 */
  constructor(origin, language) {
    this.origin = origin;
    this.language = language;
    this.example = "";
    this.jsonTypeStructure = require("./exercise_types.json");
  }

  /**
 * Create a system instruction for the chapter.
 *
 * @param {string} level - Level of the chapter.
 * @return {string} - System instruction for the chapter.
 */
  createSystemInstruction(level) {
    return `
**Your Role:**
- You are a virtual language teacher for students of ${this.origin} 
  learning ${this.language} at level ${level}.
- Your goal is to create a JSON file with one exercise 
  chapter at level ${level}.
**Your task is to generate a new chapter based on the given information**
    `;
  }

  /**
 * Generate a chapter exercise for language learning.
 *
 * @param {string} type - Type of the chapter.
 * @param {string} level - Level of the chapter.
 * @param {string} description - Description of the chapter.
 * @return {Object} - Generated chapter
 */
  async generateChapter(type, level, description) {
    this.example = this.jsonTypeStructure.chapters[type] || null;
    if (!this.example) {
      console.log("Invalid type");
      return null;
    }

    this.createSystemInstruction(level);
    const vertexAI = new VertexAI({project: process.env.GCLOUD_PROJECT,
      location: "us-central1"});
    const model = vertexAI.preview.getGenerativeModel({
      systemInstructions: `You are a virtual assistant for learning languages.
You can help users with their questions related to language learning.
You can also help users with their exercises
and provide feedback on their answers.
Do not provide any personal information or any information that is 
not related to language learning.
Be polite and helpful.`,
      model: "gemini-1.5-pro",
      generationConfig: {
        temperature: 1,
        topP: 1,
        maxOutputTokens: 2048,
        responseMimeType: "application/json",
      },
    });

    const prompt = `Generate a chapter exercise of type ${type} with 
one exercise for students of ${this.origin} 
learning ${this.language} at level ${level}. 
The chapter description and content is as follows: ${description}.
The chapter content should match the amount of the exemplary content.
The response must be in the given JSON structure.
Your response must be in the following exact format:
    
    ${JSON.stringify(this.example)}`;

    try {
      const result = await model.generateContent(prompt);
      const response = result.response.candidates.at(0)
          .content.parts.at(0).text;

      if (!response) {
        logger.log("No response from the Gemini API. Exiting.");
        return null;
      }

      try {
        const responseJson = JSON.parse(response);

        return responseJson;
      } catch (e) {
        logger.log(`Failed to parse JSON: ${e}`);
        logger.log("Response text was:", response);
        return null;
      }
    } catch (error) {
      logger.error("Error generating content:", error);
      return null;
    }
  }
}

exports.ChapterGenerator = ChapterGenerator;
